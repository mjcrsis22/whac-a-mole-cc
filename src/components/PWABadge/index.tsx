import './styles.scss';

import { useRegisterSW } from 'virtual:pwa-register/react';

function PWABadge() {
  const {
    offlineReady: [offlineReady, setOfflineReady],
    needRefresh: [needRefresh, setNeedRefresh],
    updateServiceWorker,
  } = useRegisterSW();

  function close() {
    setOfflineReady(false);
    setNeedRefresh(false);
  }

  return (
    <div className="PWABadge" role="alert" aria-labelledby="toast-message">
      {(offlineReady || needRefresh) && (
        <div className="PWABadge-toast">
          <div className="PWABadge-message">
            {offlineReady ? (
              <span id="toast-message">App ready to work offline</span>
            ) : (
              <span id="toast-message">
                New content available, click on reload button to update.
              </span>
            )}
          </div>
          <div className="PWABadge-buttons">
            {needRefresh && (
              <button
                className="button PWABadge-toast-button"
                onClick={() => updateServiceWorker(true)}
              >
                Reload
              </button>
            )}
            <button
              className="button PWABadge-toast-button"
              onClick={() => close()}
            >
              Close
            </button>
          </div>
        </div>
      )}
    </div>
  );
}

export default PWABadge;
