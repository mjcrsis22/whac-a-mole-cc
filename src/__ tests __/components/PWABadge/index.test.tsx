import { render } from '@testing-library/react';
import PWABadge from '../../../components/PWABadge';
import { useRegisterSW } from 'virtual:pwa-register/react';
import userEvent from '@testing-library/user-event';

const mSetOfflineReady = jest.fn();
const mSetNeedRefresh = jest.fn();
const mUpdateServiceWorker = jest.fn();

jest.mock(
  'virtual:pwa-register/react',
  () => ({
    useRegisterSW: jest.fn(() => ({
      offlineReady: [false, mSetOfflineReady],
      needRefresh: [false, mSetNeedRefresh],
      updateServiceWorker: mUpdateServiceWorker,
    })),
  }),
  { virtual: true },
);

const mockedUseRegisterSW = useRegisterSW as jest.Mock;

describe('PWABadge component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('Renders correctly', () => {
    const { container } = render(<PWABadge />);
    expect(container).toMatchSnapshot();
  });

  test('Renders correctly - offlineReady state', () => {
    mockedUseRegisterSW.mockReturnValue({
      offlineReady: [true, mSetOfflineReady],
      needRefresh: [false, mSetNeedRefresh],
      updateServiceWorker: mUpdateServiceWorker,
    });

    const { container } = render(<PWABadge />);
    expect(container).toMatchSnapshot();
  });

  test('Renders correctly - needRefresh state', () => {
    mockedUseRegisterSW.mockReturnValue({
      offlineReady: [false, mSetOfflineReady],
      needRefresh: [true, mSetNeedRefresh],
      updateServiceWorker: mUpdateServiceWorker,
    });

    const { container } = render(<PWABadge />);
    expect(container).toMatchSnapshot();
  });

  test('Performs reload action', async () => {
    mockedUseRegisterSW.mockReturnValue({
      offlineReady: [false, mSetOfflineReady],
      needRefresh: [true, mSetNeedRefresh],
      updateServiceWorker: mUpdateServiceWorker,
    });

    const result = render(<PWABadge />);

    const button = result.getByText('Reload', { selector: 'button' });
    await userEvent.click(button);

    expect(mUpdateServiceWorker).toHaveBeenCalledWith(true);
  });

  test('Performs close action', async () => {
    mockedUseRegisterSW.mockReturnValue({
      offlineReady: [false, mSetOfflineReady],
      needRefresh: [true, mSetNeedRefresh],
      updateServiceWorker: mUpdateServiceWorker,
    });

    const result = render(<PWABadge />);

    const button = result.getByText('Close', { selector: 'button' });
    await userEvent.click(button);

    expect(mSetOfflineReady).toHaveBeenCalledWith(false);
    expect(mSetNeedRefresh).toHaveBeenCalledWith(false);
  });
});
