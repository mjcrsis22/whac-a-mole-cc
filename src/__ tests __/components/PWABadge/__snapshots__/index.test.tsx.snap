// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PWABadge component Renders correctly - needRefresh state 1`] = `
<div>
  <div
    aria-labelledby="toast-message"
    class="PWABadge"
    role="alert"
  >
    <div
      class="PWABadge-toast"
    >
      <div
        class="PWABadge-message"
      >
        <span
          id="toast-message"
        >
          New content available, click on reload button to update.
        </span>
      </div>
      <div
        class="PWABadge-buttons"
      >
        <button
          class="button PWABadge-toast-button"
        >
          Reload
        </button>
        <button
          class="button PWABadge-toast-button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PWABadge component Renders correctly - offlineReady state 1`] = `
<div>
  <div
    aria-labelledby="toast-message"
    class="PWABadge"
    role="alert"
  >
    <div
      class="PWABadge-toast"
    >
      <div
        class="PWABadge-message"
      >
        <span
          id="toast-message"
        >
          App ready to work offline
        </span>
      </div>
      <div
        class="PWABadge-buttons"
      >
        <button
          class="button PWABadge-toast-button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PWABadge component Renders correctly 1`] = `
<div>
  <div
    aria-labelledby="toast-message"
    class="PWABadge"
    role="alert"
  />
</div>
`;
